docs/managing_deis/cache_settings.rst
:title: Customizing cache
:description: Learn how to tune custom Deis settings.

.. _cache_settings:

Customizing cache
=========================
The following settings are tunable for the :ref:`cache` component. Values are stored in etcd.

Dependencies
------------
Requires: none

Required by: :ref:`controller <controller_settings>`, :ref:`registry <registry_settings>`

Considerations: none

Settings set by cache
---------------------
The following etcd keys are set by the cache component, typically in its /bin/boot script.

================              ==============================================
setting                       description
================              ==============================================
/builtdock/cache/host              IP address of the host running cache
/builtdock/cache/port              port used by the cache service (default: 6379)
================              ==============================================

Settings used by cache
----------------------
The cache component uses no keys from etcd.

Using a custom cache image
--------------------------
You can use a custom Docker image for the cache component instead of the image
supplied with Deis:

.. code-block:: console

    $ etcdctl set /builtdock/cache/image myaccount/myimage:latest

This will pull the image from the public Docker registry. You can also pull from a private
registry:

.. code-block:: console

    $ etcdctl set /builtdock/cache/image registry.mydomain.org:5000/myaccount/myimage:latest

Be sure that your custom image functions in the same way as the `stock cache image`_ shipped with
Deis. Specifically, ensure that it sets and reads appropriate etcd keys.

.. _`stock cache image`: https://github.com/builtdock/builtdock/tree/master/cache


docs/managing_deis/logger_settings.rst
:title: Customizing logger
:description: Learn how to tune custom Deis settings.

.. _logger_settings:

Customizing logger
=========================
The following settings are tunable for the :ref:`logger` component.

Dependencies
------------
Requires: none

Required by: :ref:`controller <controller_settings>`

Considerations: must live on the same host as controller (see `#985`_)

Settings set by logger
------------------------
The following etcd keys are set by the database component, typically in its /bin/boot script.

===========================              =================================================================================
setting                                  description
===========================              =================================================================================
/builtdock/logger/host                        IP address of the host running logger
/builtdock/logger/port                        port used by the logger service (default: 514)
===========================              =================================================================================

Settings used by logger
-------------------------
The logger component uses no keys from etcd.

Using a custom logger image
---------------------------
You can use a custom Docker image for the logger component instead of the image
supplied with Deis:

.. code-block:: console

    $ etcdctl set /builtdock/logger/image myaccount/myimage:latest

This will pull the image from the public Docker registry. You can also pull from a private
registry:

.. code-block:: console

    $ etcdctl set /builtdock/logger/image registry.mydomain.org:5000/myaccount/myimage:latest

Be sure that your custom image functions in the same way as the `stock logger image`_ shipped with
Deis. Specifically, ensure that it sets and reads appropriate etcd keys.

.. _`stock logger image`: https://github.com/builtdock/builtdock/tree/master/logger
.. _`#985`: https://github.com/builtdock/deis/issues/985


CHANGELOG.md
### v0.11.0 -> v0.12.0

#### Features

 - [`5cf53e9`](http://github.com/builtdock/builtdock/commit/5cf53e90ebfc5cd328da9289c88a540b8956406a) tests: add test-nightly.sh script for CI
 - [`c3b262a`](http://github.com/builtdock/builtdock/commit/c3b262a053f853910aab4d4652b485fbbba1f9bc) tests: add end-to-end acceptance test script for vagrant
 - [`16e92ee`](http://github.com/builtdock/builtdock/commit/16e92ee919ae9f8414a270a6c2572bec719cbb74) controller: inject tag version value in environment.
 - [`49daee2`](http://github.com/builtdock/builtdock/commit/49daee2a86cad07f558700a38d10b0b74ea551c0) controller: inject tag version value in environment.
 - [`940752b`](http://github.com/builtdock/builtdock/commit/940752bb8b09749f90f250fb637597f3c7ca7bce) version: add version constant
 - [`32b3aa5`](http://github.com/builtdock/builtdock/commit/32b3aa54cab2ab74b2084d7fb92b1e814e821742) version: add -dev flag for unversioned releases
 - [`f9a736b`](http://github.com/builtdock/builtdock/commit/f9a736b5316876229878c05c92380ac7837a5369) router: custom timeouts for builder and controller
 - [`e5488e8`](http://github.com/builtdock/builtdock/commit/e5488e892dbe20fe5e698e43be13de61a5265edb) router: custom request size.
 - [`9b72b74`](http://github.com/builtdock/builtdock/commit/9b72b74d7ca178adfc251f17b94919c28c574b5a) client: add colorized logging
 - [`9bc583c`](http://github.com/builtdock/builtdock/commit/9bc583cfe94a36514a701f48410c7d55102ae0d2) tests: add CI node setup script
 - [`4ed700b`](http://github.com/builtdock/builtdock/commit/4ed700bd07cc740950857fcbabfaa09b0c94052e) controller: pipe release notes to app logs
 - [`c1c1017`](http://github.com/builtdock/builtdock/commit/c1c10172c953018a616e6a79f8526239ea7abaf5) contrib: Blacklist DigitalOcean regions.

#### Fixes

 - [`d5b419e`](http://github.com/builtdock/builtdock/commit/d5b419e055cf3446569a0892bfc979120443e04c) tests: make test-nightly.sh executable
 - [`7aa8062`](http://github.com/builtdock/builtdock/commit/7aa8062d46509379032b4c136e4d4f73e38076fe) controller: fix scale log event on git push
 - [`d6f7dd0`](http://github.com/builtdock/builtdock/commit/d6f7dd0548255d69516730db74fb7273220ef2b9) contrib/gce: fix docker storage being wiped on boot
 - [`2a8488c`](http://github.com/builtdock/builtdock/commit/2a8488ccdafb29a26a0a79f5bec2028e4811fdb4) Vagrantfile: specify AMD network to work around transfer issues
 - [`2220271`](http://github.com/builtdock/builtdock/commit/22202712e863c7888b83d61ffcdfde5f70767a55) tests: improve functional test setup logic
 - [`5bc1c80`](http://github.com/builtdock/builtdock/commit/5bc1c809b90cf0fe3e8c09b11261aa19b0f789ce) user-data: add fixed version of sed command to "make discovery-url"
 - [`27860be`](http://github.com/builtdock/builtdock/commit/27860bee20e38a3a9eceea464d383253ecd6af0c) makefile: default to 3 nodes, test for docker binary
 - [`0943823`](http://github.com/builtdock/builtdock/commit/0943823d28fd5c397e14951f7ada83c8bbd4828d) readme: address feedback, fix typos
 - [`8d08d43`](http://github.com/builtdock/builtdock/commit/8d08d430d0b65cfc12e5983b88d83c9a5a114105) user-data: restore default discovery url
 - [`d34b839`](http://github.com/builtdock/builtdock/commit/d34b8393e7a760cdb8c68126f08b280a708f7fa2) controller: properly purge domain etcd keys on app deleteion
 - [`0920b9e`](http://github.com/builtdock/builtdock/commit/0920b9ee039f8b3ad5d1db8b0fb9cd33f117252d) client: also encode keys which could be int
 - [`66333da`](http://github.com/builtdock/builtdock/commit/66333da4ab1797d5836811fb3b26841dfe4637ae) client: fix syntax error in config:list
 - [`79dd499`](http://github.com/builtdock/builtdock/commit/79dd4991fee82ad3d1b633bc7d9830592312d605) contrib/ec2: use conditional root DeviceName
 - [`99f0152`](http://github.com/builtdock/builtdock/commit/99f01525d9fe830055385876f1a35dc5bd5bcca9) controller+client: utf-8 encode only string types
 - [`8092009`](http://github.com/builtdock/builtdock/commit/8092009b40d55974fdf13e9fdb9cb2739df729aa) tests: temporary workaround to try and get jenkins passing
 - [`e3780fd`](http://github.com/builtdock/builtdock/commit/e3780fd5f8d5a1a12600a43d532e42226ccf66d9) client: use only stdout handler
 - [`8b0158e`](http://github.com/builtdock/builtdock/commit/8b0158ec2a8b4e363399e4cc66002489b91c53c9) scheduler: add gunicorn timeout for long-running methods
 - [`0cf4886`](http://github.com/builtdock/builtdock/commit/0cf488677aafc91a8828bbf611db2303e52ea434) scheduler: handle announcer timeouts to work around https://github.com/docker/docker/issues/8022
 - [`b8c9a43`](http://github.com/builtdock/builtdock/commit/b8c9a430b0dce59eb38d9a1ee603c99b9188c0a0) contrib/coreos: bump etcd peer heartbeat to 500ms
 - [`b2b8cbc`](http://github.com/builtdock/builtdock/commit/b2b8cbc5c4d168c814b075d60171aab02ef0ebbe) logger: properly parse incoming syslog messages"
 - [`230a8c6`](http://github.com/builtdock/builtdock/commit/230a8c685f8f9fefe335d1167eb008443a5bb586) client: convert client.yaml to client.json once if needed
 - [`1d20351`](http://github.com/builtdock/builtdock/commit/1d203512f9a38986679a7a91681697478a7fd223) tests: use our own fork of postgresql container for testing
 - [`362b042`](http://github.com/builtdock/builtdock/commit/362b04207430679bd3c2c57ba6ee37d4c2be4cae) builder: hardcode STACK to cedar
 - [`c15d909`](http://github.com/builtdock/builtdock/commit/c15d909c4dfa58f28b82c6f0e07b7cc730606ac9) controller: allow unicode in config:set and app logging
 - [`7df6b45`](http://github.com/builtdock/builtdock/commit/7df6b45ac384ce6b3c98a1c28ed6aa1c66dd2ab6) client: concatenate strings together
 - [`4f78bfd`](http://github.com/builtdock/builtdock/commit/4f78bfd788dc7e67d5a4537ae3058a02c8232e51) docs: update links to buildpacks included with Deis
 - [`7c9fce2`](http://github.com/builtdock/builtdock/commit/7c9fce2d83cec93cd00a1e3247e838e27bd0c49a) test: Verify apps:run output

controller/deis/settings.py
"""
Django settings for the Deis project.
"""

from __future__ import unicode_literals
import os.path
import sys
import tempfile

PROJECT_ROOT = os.path.normpath(os.path.join(os.path.dirname(__file__), '..'))

DEBUG = False
TEMPLATE_DEBUG = DEBUG

ADMINS = (
    # ('Your Name', 'your_email@example.com'),
)

MANAGERS = ADMINS

CONN_MAX_AGE = 60 * 3

# Hosts/domain names that are valid for this site; required if DEBUG is False
# See https://docs.djangoproject.com/en/1.5/ref/settings/#allowed-hosts
ALLOWED_HOSTS = ['localhost']

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# In a Windows environment this must be set to your system time zone.
TIME_ZONE = 'America/Denver'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale.
USE_L10N = True

# If you set this to False, Django will not use timezone-aware datetimes.
USE_TZ = True

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/var/www/example.com/media/"
MEDIA_ROOT = ''

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://example.com/media/", "http://media.example.com/"
MEDIA_URL = ''

# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/var/www/example.com/static/"
STATIC_ROOT = os.path.abspath(os.path.join(__file__, '..', '..', 'static'))

# URL prefix for static files.
# Example: "http://example.com/static/", "http://static.example.com/"
STATIC_URL = '/static/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = None  # @UnusedVariable

# List of callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
)

TEMPLATE_CONTEXT_PROCESSORS = (
    "django.contrib.auth.context_processors.auth",
    "django.core.context_processors.debug",
    "django.core.context_processors.i18n",
    "django.core.context_processors.media",
    "django.core.context_processors.request",
    "django.core.context_processors.static",
    "django.core.context_processors.tz",
    "django.contrib.messages.context_processors.messages",
    "allauth.account.context_processors.account",
    "deis.context_processors.site",
)

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'api.middleware.VersionMiddleware',
    # Uncomment the next line for simple clickjacking protection:
    # 'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

ROOT_URLCONF = 'deis.urls'

# Python dotted path to the WSGI application used by Django's runserver.
WSGI_APPLICATION = 'deis.wsgi.application'

TEMPLATE_DIRS = (
    # Put strings here, like "/home/html/django_templates"
    # or "C:/www/django/templates".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
    PROJECT_ROOT + '/web/templates',
)

INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.humanize',
    'django.contrib.messages',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.staticfiles',
    # Third-party apps
    'allauth',
    'allauth.account',
    'django_fsm',
    'guardian',
    'json_field',
    'gunicorn',
    'rest_framework',
    'south',
    # Deis apps
    'api',
    'web',
)

AUTHENTICATION_BACKENDS = (
    # Needed to login by username in Django admin, regardless of `allauth`
    "django.contrib.auth.backends.ModelBackend",
    "guardian.backends.ObjectPermissionBackend",
    # `allauth` specific authentication methods, such as login by e-mail
    "allauth.account.auth_backends.AuthenticationBackend",
)

ANONYMOUS_USER_ID = -1
ACCOUNT_EMAIL_REQUIRED = True
ACCOUNT_EMAIL_VERIFICATION = 'none'
ACCOUNT_LOGOUT_ON_GET = True
ACCOUNT_USERNAME_BLACKLIST = ['system']
LOGIN_REDIRECT_URL = '/dashboard/'


SOUTH_TESTS_MIGRATE = False

REST_FRAMEWORK = {
    'DEFAULT_MODEL_SERIALIZER_CLASS':
    'rest_framework.serializers.ModelSerializer',
    'DEFAULT_PERMISSION_CLASSES': (
        'rest_framework.permissions.IsAuthenticated',
    ),
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.SessionAuthentication',
    ),
    'PAGINATE_BY': 100,
}

# URLs that end with slashes are ugly
APPEND_SLASH = False

# Determine where to send syslog messages
if os.path.exists('/dev/log'):           # Linux rsyslog
    SYSLOG_ADDRESS = '/dev/log'
elif os.path.exists('/var/log/syslog'):  # Mac OS X syslog
    SYSLOG_ADDRESS = '/var/log/syslog'
else:                                    # default SysLogHandler address
    SYSLOG_ADDRESS = ('localhost', 514)

# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error when DEBUG=False.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'verbose': {
            'format': '%(levelname)s %(asctime)s %(module)s %(process)d %(thread)d %(message)s'
        },
        'simple': {
            'format': '%(levelname)s %(message)s'
        },
    },
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse'
        }
    },
    'handlers': {
        'null': {
            'level': 'DEBUG',
            'class': 'logging.NullHandler',
        },
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'simple'
        },
        'mail_admins': {
            'level': 'ERROR',
            'filters': ['require_debug_false'],
            'class': 'django.utils.log.AdminEmailHandler'
        },
        'rsyslog': {
            'class': 'logging.handlers.SysLogHandler',
            'address': SYSLOG_ADDRESS,
            'facility': 'local0',
        },
    },
    'loggers': {
        'django': {
            'handlers': ['null'],
            'level': 'INFO',
            'propagate': True,
        },
        'django.request': {
            'handlers': ['console', 'mail_admins'],
            'level': 'WARNING',
            'propagate': True,
        },
        'api': {
            'handlers': ['console', 'mail_admins', 'rsyslog'],
            'level': 'INFO',
            'propagate': True,
        },
    }
}
TEST_RUNNER = 'api.tests.SilentDjangoTestSuiteRunner'

# etcd settings
ETCD_HOST, ETCD_PORT = os.environ.get('ETCD', '127.0.0.1:4001').split(',')[0].split(':')

# default deis settings
DEIS_LOG_DIR = os.path.abspath(os.path.join(__file__, '..', '..', 'logs'))
LOG_LINES = 1000
TEMPDIR = tempfile.mkdtemp(prefix='deis')
DEFAULT_BUILD = 'builtdock/helloworld'

# security keys and auth tokens
SECRET_KEY = os.environ.get('DEIS_SECRET_KEY', 'CHANGEME_sapm$s%upvsw5l_zuy_&29rkywd^78ff(qi')
BUILDER_KEY = os.environ.get('DEIS_BUILDER_KEY', 'CHANGEME_sapm$s%upvsw5l_zuy_&29rkywd^78ff(qi')

# registry settings
REGISTRY_MODULE = 'registry.mock'
REGISTRY_URL = 'http://localhost:5000'
REGISTRY_HOST = 'localhost'
REGISTRY_PORT = 5000

# check if we can register users with `deis register`
REGISTRATION_ENABLED = True

# check if we should enable the web UI module
WEB_ENABLED = False

# default to sqlite3, but allow postgresql config through envvars
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.' + os.environ.get('DATABASE_ENGINE', 'postgresql_psycopg2'),
        'NAME': os.environ.get('DATABASE_NAME', 'deis'),
    }
}

APP_URL_REGEX = '[a-z0-9-]+'

# SECURITY: change this to allowed fqdn's to prevent host poisioning attacks
# see https://docs.djangoproject.com/en/1.5/ref/settings/#std:setting-ALLOWED_HOSTS
ALLOWED_HOSTS = ['*']

# Honor HTTPS from a trusted proxy
# see https://docs.djangoproject.com/en/1.6/ref/settings/#secure-proxy-ssl-header
SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')

# Create a file named "local_settings.py" to contain sensitive settings data
# such as database configuration, admin email, or passwords and keys. It
# should also be used for any settings which differ between development
# and production.
# The local_settings.py file should *not* be checked in to version control.
try:
    from .local_settings import *  # @UnusedWildImport # noqa
except ImportError:
    pass


# have confd_settings within container execution override all others
# including local_settings (which may end up in the container)
if os.path.exists('/templates/confd_settings.py'):
    sys.path.append('/templates')
    from confd_settings import *  # noqa
